### PR TITLE
Allow import of manual tests

### PR DESCRIPTION
**Pull request by hegerj**
Add support for importing manual test cases from Nitrate.

### CODE REVIEW

**Comment by hegerj** [comment]
The code is not very pretty and there is quite a lot of code repetetion, however I have yet to figure out how to integrate it into existing code better. Advice welcomed.
Also documentation is missing.

**Comment by hegerj** [comment]
@athwale would you please test this PR as it is and provide a feedback?

File: tmt/cli.py [hunk 1]
```diff
@@ -354,9 +354,18 @@ def create(context, name, template, force, **kwargs):
...
     help='Import disabled test cases from Nitrate as well.')
+@click.option(
+    '--manual', default=False, is_flag=True,
+    help='Import manual test cases from Nitrate.')
+@click.option(
+    '--plan', metavar='PLAN',
+    help='Import all manual test cases from Nitrate test plan.')
+@click.option(
+    '--case', metavar='CASE',
+    help='Import manual test case from Nitrate.')
```
Comment (psss) [inline tmt/cli.py:365]: This should probably say something like "Identifier of test case to be imported."

File: tmt/cli.py [hunk 2]
```diff
@@ -371,6 +380,15 @@ def import_(context, paths, makefile, nitrate, purpose, disabled, **kwargs):
                    environment, relevancy, enabled
     """
     tmt.Test._save_context(context)
+
+    if manual:
+        if not (case or plan):
+            raise tmt.utils.GeneralError(
+                "Option --case or --plan is mandatiry when using --manual.")
```
Comment (psss) [inline tmt/cli.py:387]:
```suggestion
                "Option --case or --plan is mandatory when using --manual.")
```

File: tmt/convert.py [hunk 3]
```diff
@@ -46,6 +46,116 @@ def repr_str(dumper, data):
...
+
+        os.chdir(dir_name)
+        echo("Importing the '{0}' test case.".format(dir_name))
+
+        # Test case data
+        md_content = {}
+        md_content['Setup'] = testcase.setup
+        md_content['Actions'] = testcase.action
+        md_content['Breakdown'] = testcase.breakdown
+        md_content['Expected Results'] = testcase.effect
```
Comment (psss) [inline tmt/convert.py:105]: Please, keep the naming defined in the [specification](https://tmt.readthedocs.io/en/latest/spec/tests.html#manual):

```markdown
# Setup
Prepare the env.

# Test

## Step
Do this and that.

## Expect
Check this and that.

# Cleanup
Cleanup the env.
```

Note that steps and expected results should be the second level heading.

File: tmt/convert.py [hunk 4]
```diff
@@ -46,6 +46,116 @@ def repr_str(dumper, data):
 #  Convert
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+def read_manual(plan_id, case_id, disabled):
+    """
+    Reads metadata of manual test cases from Nitrate
+    """
+    try:
+        import nitrate
```
Comment (psss) [inline tmt/convert.py:54]: We should probably use/share the `import_nitrate()` function from `export.py` and catch possible `gssapi` exceptions as it is done there.

File: tmt/convert.py [hunk 5]
```diff
@@ -46,6 +46,116 @@ def repr_str(dumper, data):
...
+        else:
+            case_ids = [int(case_id)]
+    except ValueError:
+        raise ConvertError('Value of plan/case option must be an integer')
+
+    # Create directory to store manual tests in
+    old_cwd = os.getcwd()
+    os.chdir(tree.root)
+    try:
+        os.mkdir('Manual')
```
Comment (psss) [inline tmt/convert.py:78]: Let's use lowercase to keep consistency with other templates.

File: tmt/convert.py [hunk 6]
```diff
@@ -46,6 +46,116 @@ def repr_str(dumper, data):
...
+        md_content['Actions'] = testcase.action
+        md_content['Breakdown'] = testcase.breakdown
+        md_content['Expected Results'] = testcase.effect
+
+        # Test case metadata
+        data = {}
+        data['manual'] = True
+        data['test'] = 'full.md'
+
+        # Relevancy
```
Comment (psss) [inline tmt/convert.py:112]: It would be good to import just all nitrate attributes the same way as it is done in the `read` function. I'd suggest to create a new function for this shared stuff.

File: tmt/convert.py [hunk 7]
```diff
@@ -46,6 +46,116 @@ def repr_str(dumper, data):
...
+        echo(style('enabled: ', fg='green') + str(data['enabled']))
+
+        write_markdown(os.getcwd() + '/full.md', md_content)
+        write(os.getcwd() + '/main.fmf', data)
+        os.chdir('..')
+
+    os.chdir(old_cwd)
+
+
+def write_markdown(path, content):
```
Comment (psss) [inline tmt/convert.py:141]: Would it make sense to use a [python markdown module](https://pypi.org/project/Markdown/) for handling Markdown? If/when we are able to export manual cases I guess we will need it anyway.

File: tmt/convert.py (hunk 5 above)
Comment (hegerj) [reply tmt/convert.py:78]: I chose uppercases to keep consistency with other dist git test directories (Regression, Sanity...)

File: tmt/convert.py (hunk 7 above)
Comment (hegerj) [reply tmt/convert.py:141]: From what I could find out the markdown module it is only/mostly used to markdown -> html conversion rather than plaintext -> markdown or even html -> markdown which would also be  useful.
If you know how to use it better please let me know.

File: tmt/convert.py (hunk 3 above)
Comment (hegerj) [reply tmt/convert.py:105]: As far as I can see there is nothing in specification about mapping existing format of manual tests to this new format.
The issue is that new specification has only 3 main sections (Setup, Test, Cleanup) to 4 sections of old one (Setup, Action, Breakdown, Expected result).
Mapping Setup-Setup is obvious however the rest is quite problematic in my opinion.
I could map old Action and Expected result under new Test section (Action would be second level Step and Expected Result would be second level Expect), but I am not sure it would work for most tests.
Cleanup would be empty.
I have no idea what to do with Breakdown though.

Any ideas?

File: tmt/cli.py (hunk 1 above)
Comment (hegerj) [reply tmt/cli.py]: fixed in 440e5c1

File: tmt/cli.py (hunk 2 above)
Comment (hegerj) [reply tmt/cli.py]: fixed in 440e5c1

File: tmt/convert.py (hunk 4 above)
Comment (hegerj) [reply tmt/convert.py]: fixed in 440e5c1
Added returning of the module to the import_nitrate() method so it can be used outside of its scope.

File: tmt/convert.py (hunk 6 above)
Comment (hegerj) [reply tmt/convert.py]: Addressed in fb5ef79

File: tmt/convert.py (hunk 3 above)
Comment (psss) [reply tmt/convert.py:102]: I'd suggest the following mapping:

* setup ... setup
* action ... test/step
* expected result ... test/expect
* breakdown ... cleanup

We don't have to cover all possible corner cases. Users can adjust content after import.

File: tmt/convert.py (hunk 5 above)
Comment (psss) [reply tmt/convert.py:75]: I see. I believe in the near future we will/should get rid of this old classification which IMO does not make any sense. But those old directories will be with us for some time. And users are expected to rename / move imported tests around as necessary. So I leave the naming up to you.

File: tmt/convert.py (hunk 7 above)
Comment (psss) [reply tmt/convert.py:116]: I tried to import an example test and the result looks like this:

    # Setup
    <p>Do this and that to setup the environment.</p>

    # Actions
    <p>Step one.</p>
    <p>Step two.</p>
    <p>Step three.</p>

    # Breakdown
    <p>This is a breakdown.</p>

    # Expected Results
    <p>Expect one.</p>
    <p>Expect two</p>
    <p>Expect three.</p>

I guess the `markdown` module could/would take care of handling the html code and convert paragraph tags into new lines?

File: tmt/convert.py (hunk 3 above)
Comment (hegerj) [reply tmt/convert.py]: Changed in d65b862
However in cases I tried to import the Cleanup setion (original Breakdown) does not contain cleanup instruction as I feared. It will have to be manually changed or it won't make much sense.

File: tmt/convert.py (hunk 7 above)
Comment (hegerj) [reply tmt/convert.py:116]: As I said in previous comment I do not believe this is how `markdown` module works. As far as I know it is intended for converting markdown->html not the other way around.

I found a way how to *sort of* convert html to md however it uses non-standart library (html2text) and the results are always the best.

Comment (psss) [reply tmt/convert.py:116]: I see, thanks for clarification. I think it adding `python3-html2text` to the dependencies should be OK. It's just a small package and it does not bring many other deps. It would be good to prevent having many `<p>` and other html tags in the imported content, even if the conversion is not always ideal.

File: tmt/convert.py (hunk 3 above)
Comment (psss) [reply tmt/convert.py]: Thanks for fixing this.

**Comment by psss** [comment]
Thanks for the updates, @hegerj. I would just propose to use the above-mentioned `html2text` to get rid of the html tags and squash the pull request commits. Otherwise looks good to me.

@Athwale and/or @kkaarreell, could you please check the proposed functionality and provide feedback if the manual test case import works as you'd expected? Thanks.

File: tmt/convert.py (hunk 7 above)
Comment (psss) [reply tmt/convert.py:132]: Looks good to me after last changes. Just rebase and squash.

**Comment by Athwale** [comment]
> @Athwale and/or @kkaarreell, could you please check the proposed functionality and provide feedback if the manual test case import works as you'd expected? Thanks.

@psss I was on PTO last week. Not sure if there is anything to add now, but I checked the example convert and it look fine to me.

**Comment by psss** [comment]
@Athwale, good, thanks for the check. Added just a couple of minor style adjustments and going to merge.
